### PR TITLE
Add coverage for katello-ca certificate refresh

### PR DESCRIPTION
**Pull request by shubhamsg199**
Adding coverage for katello-ca certificate refresh

### CODE REVIEW

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8063
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========= 2 failed, 16 deselected, 138 warnings in 1507.44s (0:25:07) ==========
```

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8066
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========= 2 failed, 16 deselected, 132 warnings in 1218.73s (0:20:18) ==========
```

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8069
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========== 16 deselected, 23 warnings, 2 errors in 203.04s (0:03:23) ===========
```

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8070
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========= 2 passed, 16 deselected, 155 warnings in 1335.22s (0:22:15) ==========
```

File: tests/foreman/api/test_registration.py [hunk 1]
```diff
@@ -365,3 +365,88 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    org = module_sca_manifest_org
+    result = rhel_contenthost.api_register(
+        module_target_sat,
+        organization=org,
+        activation_keys=[module_activation_key.name],
+    )
+    assert result.status == 0, f'Failed to register host: {result.stderr}'
+
+    # /etc/rhsm/ca/katello-server-ca.pem
+    ca_file_before = len(str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem')))
```
Comment (shweta83) [inline tests/foreman/api/test_registration.py:408]: Can you please move the cert path to variable and use that instead in all the places?

Comment (shweta83) [inline tests/foreman/api/test_registration.py:408]:
```suggestion
    ca_cert_path = len(str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem')))
```

File: tests/foreman/api/test_registration.py [hunk 2]
```diff
@@ -365,3 +365,88 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+
+    # corrupt the certificate file
+    corrupt_data = gen_string('alphanumeric')
+    result = rhel_contenthost.execute(
+        f'sed -i "$ a {corrupt_data}" /etc/rhsm/ca/katello-server-ca.pem'
+    )
+    assert result.status == 0
+    ca_file_after = len(str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem')))
+    assert ca_file_before != ca_file_after
```
Comment (shweta83) [inline tests/foreman/api/test_registration.py:417]:
```suggestion
    assert ca_cert_file != corrupted_crt_file
```

File: tests/foreman/api/test_registration.py [hunk 3]
```diff
@@ -365,3 +365,88 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+            'inputs': {
+                'url': f'https://{module_target_sat.hostname}/unattended/public/foreman_ca_refresh'
+            },
+        },
+    )
+    module_target_sat.wait_for_tasks(
+        f'resource_type = JobInvocation and resource_id = {job["id"]}', poll_timeout=1000
+    )
+    result = module_target_sat.api.JobInvocation(id=job['id']).read()
+    assert result.pending == 0
```
Comment (shweta83) [inline tests/foreman/api/test_registration.py:444]: IMO any of the 2 assertions would work. Asserting pending is not needed I guess.

File: tests/foreman/api/test_registration.py [hunk 4]
```diff
@@ -365,3 +365,88 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    assert result.succeeded == 1
+    assert result.status_label == 'succeeded'
+
+    # check if the certificate file is refreshed
+    ca_file_after_refresh = len(
+        str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem'))
+    )
+    assert ca_file_before == ca_file_after_refresh
```
Comment (shweta83) [inline tests/foreman/api/test_registration.py:452]:
```suggestion
    assert ca_cert_file == ca_file_after_refresh
```

**Review by shweta83 — CHANGES_REQUESTED** [review]

File: tests/foreman/api/test_registration.py (hunk 1 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

File: tests/foreman/api/test_registration.py [hunk 5]
```diff
@@ -365,3 +365,88 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    # /etc/rhsm/ca/katello-server-ca.pem
+    ca_file_before = len(str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem')))
+
+    # corrupt the certificate file
+    corrupt_data = gen_string('alphanumeric')
+    result = rhel_contenthost.execute(
+        f'sed -i "$ a {corrupt_data}" /etc/rhsm/ca/katello-server-ca.pem'
+    )
+    assert result.status == 0
+    ca_file_after = len(str(rhel_contenthost.execute('cat /etc/rhsm/ca/katello-server-ca.pem')))
```
[the remark this answers is not in the document]
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

File: tests/foreman/api/test_registration.py (hunk 2 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

File: tests/foreman/api/test_registration.py (hunk 3 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

File: tests/foreman/api/test_registration.py (hunk 4 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: updated

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8096
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========== 16 deselected, 23 warnings, 2 errors in 202.19s (0:03:22) ===========
```

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8098
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========= 2 passed, 16 deselected, 651 warnings in 7489.38s (2:04:49) ==========
```

File: tests/foreman/api/test_registration.py [hunk 6]
```diff
@@ -365,3 +365,83 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
     # verify status code when registrationCommand fails to register on host
     assert result.status == 1
     assert 'Couldn\'t find activation key' in result.stderr
+
+
+@pytest.mark.parametrize('job_type', ['ansible', 'ssh'])
+@pytest.mark.rhel_ver_match('8')
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_registration.py:371]:
```suggestion
@pytest.mark.rhel_ver_list([settings.content_host.default_rhel_version])
```

File: tests/foreman/api/test_registration.py [hunk 7]
```diff
@@ -365,3 +365,83 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    """
+    org = module_sca_manifest_org
+    result = rhel_contenthost.api_register(
+        module_target_sat,
+        organization=org,
+        activation_keys=[module_activation_key.name],
+    )
+    assert result.status == 0, f'Failed to register host: {result.stderr}'
+    katello_ca_crt_path = '/etc/rhsm/ca/katello-server-ca.pem'
+    # /etc/rhsm/ca/katello-server-ca.pem
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_registration.py:407]: this comment looks unrelated, can we remove it?

**Review by Gauravtalreja1 — COMMENTED** [review]
ACK, with minor suggestions

File: tests/foreman/api/test_registration.py (hunk 6 above)
Comment (shweta83) [reply tests/foreman/api/test_registration.py:371]: In my opinion, we should test on all rhel versions. wdyt @Gauravtalreja1 @shubhamsg199 ?

File: tests/foreman/api/test_registration.py [hunk 8]
```diff
@@ -365,3 +365,83 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    ca_cert_file = len(str(rhel_contenthost.execute(f'cat {katello_ca_crt_path}')))
+
+    # corrupt the certificate file
+    corrupt_data = gen_string('alphanumeric')
+    result = rhel_contenthost.execute(f'sed -i "$ a {corrupt_data}" {katello_ca_crt_path}')
```
[the remark this answers is not in the document]
Comment (shweta83) [reply tests/foreman/api/test_registration.py:412]: using a variable looks more neat here.

File: tests/foreman/api/test_registration.py (hunk 7 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: Removed

File: tests/foreman/api/test_registration.py [hunk 9]
```diff
@@ -365,3 +365,83 @@ def test_negative_verify_bash_exit_status_failing_host_registration(
...
+    assert result.status == 0
+    corrupted_ca_crt_file = len(str(rhel_contenthost.execute(f'cat {katello_ca_crt_path}')))
+    assert ca_cert_file != corrupted_ca_crt_file
+
+    # run the rex job to refresh the CA certificate
+    template_name = (
+        "Download and execute a script" if job_type == "ansible" else "Download and run a script"
+    )
```
[the remark this answers is not in the document]
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: I would like to keep as it is in more readable format

File: tests/foreman/api/test_registration.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py:411]: Yes, I prefer var over this

File: tests/foreman/api/test_registration.py (hunk 6 above)
Comment (shubhamsg199) [reply tests/foreman/api/test_registration.py]: I believe from functionality perspective it would be same for all rhel versions

**Comment by shubhamsg199** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8101
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_registration.py -k test_positive_katello_ca_crt_refresh --external-logging
Test Result : ========= 2 passed, 16 deselected, 164 warnings in 1688.89s (0:28:08) ==========
```

**Review by Gauravtalreja1 — APPROVED** [review]

**Review by shweta83 — APPROVED** [review]
Ack
